tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php

```diff
@@ -85,4 +85,255 @@ public function removeInvisibleNodesRemovesNodesWithDisplayNone($displayNone)
+     *
+     * @dataProvider nonMatchedClassesDataProvider
+     */
+    public function removeRedundantClassesRemovesClassAttributesContainingNoClassesToKeep($html, array $classesToKeep)
+    {
+        $subject = HtmlPruner::fromHtml('<html>' . $html . '</html>');
+
+        $subject->removeRedundantClasses($classesToKeep);
+
+        self::assertNotContains('class', $subject->render());
+    }
+
+    /**
+     * @return (string|string[])[][]
+     */
+    public function classesDataProvider()
+    {
+        $nonMatchedClassesData = $this->nonMatchedClassesDataProvider();
+        $matchedClassesData = [
+            '1 attribute, 1 class, that class to be kept' => [
+                'html' => '<p class="foo">hello</p>',
+                'classes to keep' => ['foo'],
+            ],
+            '2 attributes, 1 different class each, both classes to be kept' => [
+                'html' => '<p class="foo">hello</p><p class="bar">world</p>',
+                'classes to keep' => ['foo', 'bar'],
+            ],
+            '2 attributes, 1 different class each, 1st class to be kept' => [
+                'html' => '<p class="foo">hello</p><p class="bar">world</p>',
+                'classes to keep' => ['foo'],
+                'classes expected to be removed' => ['bar'],
+            ],
+            '2 attributes, 1 different class each, 2nd class to be kept' => [
+                'html' => '<p class="foo">hello</p><p class="bar">world</p>',
+                'classes to keep' => ['bar'],
+                'classes expected to be removed' => ['foo'],
+            ],
+            '2 attributes, same 1 class each, that class to be kept' => [
+                'html' => '<p class="foo">hello</p><p class="foo">world</p>',
+                'classes to keep' => ['foo'],
+            ],
+            '1 attribute, 2 classes, both to be kept' => [
+                'html' => '<p class="foo bar">hello</p>',
+                'classes to keep' => ['foo', 'bar'],
+            ],
+            'first class in attribute is to be removed' => [
+                'html' => '<p class="foo bar baz">hello</p>',
+                'classes to keep' => ['bar', 'baz'],
+                'classes expected to be removed' => ['foo'],
+            ],
+            'middle class in attribute is to be removed' => [
+                'html' => '<p class="foo bar baz">hello</p>',
+                'classes to keep' => ['foo', 'baz'],
+                'classes expected to be removed' => ['bar'],
+            ],
+            'last class in attribute is to be removed' => [
+                'html' => '<p class="foo bar baz">hello</p>',
+                'classes to keep' => ['foo', 'bar'],
+                'classes expected to be removed' => ['baz'],
+            ],
+        ];
+        $matchedClassesWithExtraWhitespaceData = $this->matchedClassesWithExtraWhitespaceDataProvider();
+
+        return $nonMatchedClassesData + $matchedClassesData + $matchedClassesWithExtraWhitespaceData;
```

Comment:
Instead of having a data provider with a different kinds of data sets, I propose we have three separate tests that use each data subset and focus on one aspect of the behavior.

Reply:
OK, I'll look into reworking it...